src/qobj/arithmetic_and_attributes.jl

```diff
@@ -514,18 +514,38 @@ Quantum Object:   type=Operator   dims=[2]   size=(2, 2)   ishermitian=true
 ```
 """
 function ptrace(QO::QuantumObject{<:AbstractArray,KetQuantumObject}, sel::Union{AbstractVector{Int},Tuple})
-    length(QO.dims) == 1 && return QO
+    ns = length(sel)
+    if ns == 0
+        return tr(QO * QO')
+    else
+        nd = length(QO.dims)
+        !all(nd .>= sel .> 0) && throw(
+            ArgumentError("Invalid indices in `sel`: $(sel), the given QuantumObject only have $(nd) sub-systems"),
+        )
+        ns != length(unique(sel)) && throw(ArgumentError("Duplicate selection indices in `sel`: $(sel)"))
+        nd == 1 && return QO * QO'  # ptrace should always return Operator
+    end
 
-    ρtr, dkeep = _ptrace_ket(QO.data, QO.dims, SVector(sel))
+    ρtr, dkeep = _ptrace_ket(QO.data, QO.dims, sort(SVector(sel)))
     return QuantumObject(ρtr, type = Operator, dims = dkeep)
 end
 
 ptrace(QO::QuantumObject{<:AbstractArray,BraQuantumObject}, sel::Union{AbstractVector{Int},Tuple}) = ptrace(QO', sel)
 
 function ptrace(QO::QuantumObject{<:AbstractArray,OperatorQuantumObject}, sel::Union{AbstractVector{Int},Tuple})
-    length(QO.dims) == 1 && return QO
+    ns = length(sel)
+    if ns == 0
+        return tr(QO)
+    else
+        nd = length(QO.dims)
+        !all(nd .>= sel .> 0) && throw(
```

Comment:
Same here.

Reply:
okay, I changed 
`!all(nd .>= sel .> 0)`
to
`(any(>(nd), sel) || any(<(0), sel))`